src/torchpme/potentials/potential_dipole.py

```diff
@@ -0,0 +1,134 @@
+            0
+        ) - 3.0 * r_outer / (r_mag**5).unsqueeze(-1)
+
+    @torch.jit.export
+    def sr_from_dist(self, vector: torch.Tensor) -> torch.Tensor:
+        r"""TODO: Add docstring"""
+        if self.smearing is None:
+            raise ValueError(
+                "Cannot compute range-separated potential when `smearing` is not specified."
+            )
+        if self.exclusion_radius is None:
+            return self.from_dist(vector) - self.lr_from_dist(vector)
+        return -self.lr_from_dist(vector) * self.f_cutoff(vector)
+
+    @torch.jit.export
+    def lr_from_dist(self, vector: torch.Tensor) -> torch.Tensor:
+        r"""TODO: Add docstring"""
+        if self.smearing is None:
+            raise ValueError(
+                "Cannot compute long-range contribution without specifying `smearing`."
+            )
+        alpha = 1 / (2 * self.smearing**2)
+        r_mag = torch.norm(vector, dim=1, keepdim=True)
+        scalar_potential = (
+            1.0 / (r_mag**3)
+            - torch.erfc(torch.sqrt(alpha) * r_mag) / r_mag**3
+            - 2 * torch.sqrt(alpha / torch.pi) * torch.exp(-alpha * r_mag**2) / r_mag**2
+        )
+        r_outer = torch.einsum("bi,bj->bij", vector, vector)
+        return scalar_potential.unsqueeze(-1) * torch.eye(3).to(r_outer).unsqueeze(
+            0
+        ) - r_outer * (
+            3.0 / (r_mag**5)
+            - 3.0 * torch.erfc(torch.sqrt(alpha) * r_mag) / r_mag**5
+            - 2
+            * torch.sqrt(alpha / torch.pi)
+            * (2 * alpha + 3 / r_mag**2)
+            * torch.exp(-alpha * r_mag**2)
+            / r_mag**2
+        ).unsqueeze(-1)
+
+    def lr_from_k_sq(self, k_sq: torch.Tensor) -> torch.Tensor:
+        r"""TODO: Add docstring"""
+        if self.smearing is None:
+            raise ValueError(
+                "Cannot compute long-range kernel without specifying `smearing`."
+            )
+
+        # avoid NaNs in backward, see
+        # https://github.com/jax-ml/jax/issues/1052
+        # https://github.com/tensorflow/probability/blob/main/discussion/where-nan.pdf
+        masked = torch.where(k_sq == 0, 1.0, k_sq)
+        return torch.where(
+            k_sq == 0,
+            0.0,
+            4 * torch.pi * torch.exp(-0.5 * self.smearing**2 * masked) / masked,
+        )
+
+    def self_contribution(self) -> torch.Tensor:
+        if self.smearing is None:
+            raise ValueError(
+                "Cannot compute long-range contribution without specifying `smearing`."
+            )
+        alpha = 1 / (2 * self.smearing**2)
+        return 2 * torch.pi / 3 * torch.sqrt((alpha / torch.pi) ** 3)
+
+    def background_correction(self) -> torch.Tensor:
+        if self.smearing is None:
+            raise ValueError(
+                "Cannot compute background correction without specifying `smearing`."
+            )
+        return self.smearing * 0.0
```

Comment:
Always 0?

```suggestion
        return 0.0
```

then we can even remove the value error?

Reply:
Here, it’s done this way to stay consistent with the pipeline. This ensures that 0.0 is returned as a `torch.tensor` with the correct device and dtype; otherwise, we would just return a float number.

Reply:
In general, I have no idea whether it should always be zero or not. One would need to perform a theoretical derivation to determine whether dipole neutrality is a crucial aspect for the dipole-Ewald calculator